### PR TITLE
Support kubernetes deployment type

### PR DESCRIPTION
**Pull request by osterman**
## what
* introduce support for `Deployment` type
* list deployments on circleci deploy

## why
* deployments eliminate the need for us to use the `SERIAL` variable to manage generations of a `ReplicationController`. Behind the scenes kubernetes now does this for us.
* deployments make it easier to rollback to any previous generation
* deployments won't clobber each other; if a deployment is currently underway and another deployment is triggered, the latest one will win and current deployment underway aborted
* deployments history makes it easy to review what changes to a cluster have been made

## who
@darend 

## related issues
https://github.com/kubernetes/kubernetes/issues/25554

DEV-4784

### CODE REVIEW

**Comment by darend** [comment]
> deployments won't clobber each other; if a deployment is currently underway and another deployment is triggered, the latest one will win and current deployment underway aborte

From k8s http://kubernetes.io/docs/user-guide/deployments/#multiple-updates:
> for example, suppose you create a Deployment to create 5 replicas of nginx:1.7.9, but then update the Deployment to create 5 replicas of nginx:1.9.1, when only 3 replicas of nginx:1.7.9 had been created. In that case, Deployment will immediately start killing the 3 nginx:1.7.9 Pods that it had created, and will start creating nginx:1.9.1 Pods. It will not wait for 5 replicas of nginx:1.7.9 to be created before changing course.

GENIUS

**Comment by darend** [comment]
> deployments make it easier to rollback to any previous generation

From http://kubernetes.io/docs/user-guide/deployments/#rolling-back-a-deployment

> Now we’ve decided to undo the current rollout and rollback to the previous revision:
$ kubectl rollout undo deployment/nginx-deployment
deployment "nginx-deployment" rolled back

🎉

File: modules/Makefile.kubernetes [hunk 1]
```diff
@@ -112,10 +133,15 @@ kubernetes\:deploy-service:
 ## Deploy controller and service; notify datadog
 kubernetes\:deploy:
 	$(NOTIFY_STARTING)
+	@[ ! -f "$(KUBERNETES_RESOURCE_PATH)/$(KUBERNETES_APP)-deployment.yml" ] || $(SELF) kubernetes:deploy-deployment || $(NOTIFY_FAILURE)
```
Comment (darend) [inline modules/Makefile.kubernetes]: My understanding if the `deployment.yml` replaces the `controller.yml`

File: modules/Makefile.kubernetes [hunk 2]
```diff
@@ -112,10 +133,15 @@ kubernetes\:deploy-service:
...
 	@[ ! -f "$(KUBERNETES_RESOURCE_PATH)/$(KUBERNETES_APP)-controller.yml" ] || $(SELF) kubernetes:deploy-controller || $(NOTIFY_FAILURE)
 	@[ ! -f "$(KUBERNETES_RESOURCE_PATH)/$(KUBERNETES_APP)-service.yml" ] || $(SELF) kubernetes:deploy-service || $(NOTIFY_FAILURE)
 	$(NOTIFY_SUCCESS)
 
+## List deployments
```
Comment (darend) [inline modules/Makefile.kubernetes]: would be nice to have a rollback target, We could do it in a separate PR if you prefer

File: modules/Makefile.kubernetes [hunk 3]
```diff
@@ -85,13 +85,34 @@ kubernetes\:replace-controller:
...
+# (private) Create a new deployment
+kubernetes\:create-deployment:
+	$(call kubectl_create,deployment)
+
+# (private) Delete a deploymentt
+kubernetes\:delete-deployment:
+	$(call kubectl_delete,deployment)
+
+# (private) Update existing controller using rolling-update or create a new controller; do not notify datadog
+kubernetes\:deploy-deployment:
```
Comment (darend) [inline modules/Makefile.kubernetes]: sorry but I cant help but suggest a new name, `perform-deployment`?

File: modules/Makefile.kubernetes (hunk 1 above)
Comment (osterman) [inline modules/Makefile.kubernetes]: Both resource types are supported. E.g. a `DaemonSet` is a type of controller, but a `Deployment` resource type creates `ReplicationController` resources. A `Deployment` type can only create `ReplicationController` resources and nothing else.

File: modules/Makefile.kubernetes (hunk 2 above)
Comment (osterman) [inline modules/Makefile.kubernetes]: I agree. No strong preference either way.

File: modules/Makefile.kubernetes (hunk 1 above)
Comment (darend) [inline modules/Makefile.kubernetes]: Ok I understand now

File: modules/Makefile.kubernetes (hunk 3 above)
Comment (osterman) [inline modules/Makefile.kubernetes]: Okay, can see your point. I think it depends on the angle of approach.

We have a select few verbs that we use like `create`, `delete, `replace`, and `deploy`. 

The `deploy` verb is one that does not conflict with `kubectl` actions and was chosen to mean "take the best course of action in order to accomplish some task using `kubectl` operations". 
e.g.

`kubernetes:deploy-controller`: figures out if the `controller` resource exists. if it exists, then it does a `rolling-update`, otherwise it does a `create`.

`kubernetes:deploy-service`: figures out if the `service` resource has changed or does not exist. if 
it has changed or does not exist, then it does a `kubernetes:replace` (which calls `delete` then `create`), otherwise it does nothing.

`kubernetes:deploy-deployment`: figures out if the `deployment` resource exists. if it exists, then it updates it by calling `apply` on the resource, otherwise it calls `create` instantiate a new one.

Lastly, there's the `kubernetes:deploy` target that figures out which of the `deploy-<something>` targets are applicable and calls them in the appropriate order.

I think `kubernetes:deploy` is all that a a user should care about which is why the other targets are "private" (as in not available in `make help`), but named consistently for maintainability of the code. 

It's a lose/lose. Feelings won't be hurt either way. =)

Comment (darend) [inline modules/Makefile.kubernetes]: Keep it as you have, ie `deploy-deployment`. As you say it does work because `deployment` is a type here

**Comment by osterman** [comment]
@darend ready for review

File: modules/Makefile.circleci [hunk 4]
```diff
@@ -8,9 +8,14 @@ BUILD ?= $(CIRCLE_BRANCH)-$(CIRCLE_BUILD_NUM)
 endif
 
 COMMIT ?= $(CIRCLE_SHA1)
+COMMIT_LOG ?= $(shell git log --format=oneline --pretty=format:'%s' -n 1 $(CIRCLE_SHA1))
```
Comment (darend) [inline modules/Makefile.circleci]: How are these used?

File: modules/Makefile.kubernetes [hunk 5]
```diff
@@ -20,11 +20,13 @@ export SERIAL
 
 define kubectl_create
 	@echo -e "INFO: Creating $(KUBERNETES_APP) $(1) on cluster $(call yellow,$(CLUSTER_NAMESPACE))..."
+	@[ -z "$(DEBUG)" ] && (envsubst < "$(KUBERNETES_RESOURCE_PATH)/$(KUBERNETES_APP)-$(1).yml" | envsubst)
```
Comment (darend) [inline modules/Makefile.kubernetes]: Its slightly annoying that this means if you want to make changes you need to do it in two places. 

What if you instead made `envsubst` a variable and it dynamically added the `@` if `DEBUG` was false? Then you wouldnt need the duplicate line, and you would get the full output of the kubectl line

File: modules/Makefile.kubernetes [hunk 6]
```diff
@@ -85,18 +87,47 @@ kubernetes\:replace-controller:
 kubernetes\:replace:
 	@$(SELF) kubernetes:replace-service kubernetes:replace-controller
 
+# (private) Create a new deployment
+kubernetes\:create-deployment:
```
Comment (darend) [inline modules/Makefile.kubernetes]: When would we use this vs `apply-deployment`?

The k8s docs are unclear. The example they provide shows a create for the initial deployment, followed up apply for subsequent ones.

http://kubernetes.io/docs/user-guide/deployments/#creating-a-deployment

File: modules/Makefile.circleci (hunk 4 above)
Comment (osterman) [inline modules/Makefile.circleci]: It's so we can reference this in the annotations when doing deployments. 

e.g. 
image

Not currently necessary since I removed the annotation block from the resource file, but it will hopefully be available soon given the rate they iterate & release kubernetes. They are actively working on deployments & annotations (per @janetkuo).

File: modules/Makefile.kubernetes (hunk 5 above)
Comment (osterman) [inline modules/Makefile.kubernetes]: Okay, you mean make a Makefile macro that does both? I'd be okay with that.

File: modules/Makefile.kubernetes (hunk 6 above)
Comment (osterman) [inline modules/Makefile.kubernetes]: At this point, we would never use it. I should remove it.

File: modules/Makefile.circleci (hunk 4 above)
Comment (darend) [inline modules/Makefile.circleci]: Ah nice

**Comment by darend** [comment]
+2
